pkg/trtl/honu.go

```diff
@@ -108,8 +112,179 @@ func (h *HonuService) Delete(ctx context.Context, in *pb.DeleteRequest) (out *pb
+	if in.Options != nil {
+		opts = in.Options
+	} else {
+		// Create empty options
+		opts = &pb.Options{}
+	}
+
+	// Load defaults into options
+	if opts.PageSize == 0 {
+		opts.PageSize = 100
+	}
+
+	// Compute the actual starting prefix as the namespace plus the key
+	var prefix []byte
+	if in.Namespace != "" {
+		prefix = prepend(in.Namespace, in.Prefix)
+	} else {
+		prefix = prepend("default", in.Prefix)
+	}
+
+	// If a page cursor is provided load it, otherwise create the cursor for iteration
+	cursor := &internal.PageCursor{}
+	if opts.PageToken != "" {
+		if err = cursor.Load(opts.PageToken); err != nil {
+			log.Warn().Err(err).Msg("invalid page token on iter request")
+			return nil, status.Error(codes.InvalidArgument, "invalid page token")
+		}
+
+		// Validate the request has not changed
+		if cursor.PageSize != opts.PageSize {
+			log.Debug().Int32("cursor", cursor.PageSize).Int32("opts", opts.PageSize).Msg("invalid iter request: mismatched page size")
+			return nil, status.Error(codes.InvalidArgument, "page size cannot change between requests")
+		}
+
+		if !bytes.HasPrefix(cursor.NextKey, prefix) {
+			log.Debug().Msg("invalid iter request: mismatched prefix")
+			return nil, status.Error(codes.InvalidArgument, "prefix and namespace cannot change between requests")
+		}
+
+	} else {
+		// Create a new cursor
+		cursor.PageSize = opts.PageSize
+	}
+
+	// Create response
+	out = &pb.IterReply{
+		Values: make([]*pb.KVPair, 0, cursor.PageSize),
+	}
+
+	// TODO: in order to support more complex iteration such as jumping to the next key
+	// in the page, honu needs to offer better support for leveldb iteration options.
+	// Until this is implemented, we just iterate over the prefix until we get to the
+	// start of the next page, which is extremely inefficient, especially for large
+	// datasets. [Create a story for implementing iter.Seek() in Honu]
+	var iter iterator.Iterator
+	if iter, err = h.db.Iter(prefix); err != nil {
+		log.Error().Err(err).Str("namespace", in.Namespace).Msg("could not create honu iterator")
+		return nil, status.Errorf(codes.FailedPrecondition, "could not create iterator: %s", err)
+	}
+	defer iter.Release()
+
+	for iter.Next() {
+		// Determine if we need to seek to the next page or not
+		key := iter.Key()
+		if len(cursor.NextKey) > 0 {
+			// We need to seek since there is a page token
+			// If the current key is lexicographically before the next key, then we need
+			// to continue seeking. Note that we cannot use equality here because the
+			// next key may have been deleted between requests, which means we'd seek to
+			// the end of the iteration without returning the page. Unfortunately, the
+			// lexicographic ordering that we're computing is heavily dependent on the
+			// underlying representation, so I'm just guessing with leveldb for now.
+			// TODO: this needs to be replaced with honu Seek!
+			if comparer.DefaultComparer.Compare(key, cursor.NextKey) < 0 {
+				continue
+			}
+		}
```

Comment:
Note the TODOs and the comments; probably this is a story that will require a modification to both honu and trtl. We should also look at the synchronization and configuration issues I mentioned exist in Honu as well. Alternatively we could simply remove Honu as a dependency and shift some of that code to trtl if we wanted to leave Honu in a state to evolve by itself. Some discussion required.